### PR TITLE
feat: deterministic task assignment

### PR DESCRIPTION
**Pull request by bajtos**
- **feat!: rename INVALID_TASK to TASK_NOT_IN_ROUND**
- **feat: add startEpoch to RoundDetails type**
- **feat: add DRAND client**
- **feat: deterministic task assignment**

For each station id, calculate the list of tasks assigned to this station by our deterministic task assignment algorithm. Reject measurements for other tasks.

_The pull request may be easier to review commit-by-commit._

Links:
- https://github.com/filecoin-station/spark/issues/38
- https://github.com/filecoin-station/spark/issues/27
- https://github.com/filecoin-station/spark/pull/85
- https://github.com/filecoin-station/spark-evaluate/pull/287
- https://github.com/filecoin-station/spark-evaluate/pull/295

### CODE REVIEW

File: test/evaluate.js [hunk 1]
```diff
@@ -616,12 +633,55 @@ describe('fraud detection', () => {
...
 
-    await runFraudDetection(1n, measurements, sparkRoundDetails)
+    await runFraudDetection(1n, measurements, sparkRoundDetails, logger)
     assert.deepStrictEqual(
       measurements.map(m => m.fraudAssessment),
       ['OK', 'IPNI_NOT_QUERIED']
     )
   })
+
+  it('rejects tasks not allowed for the node by the tasking algorithm', async () => {
```
Comment (juliangruber) [inline test/evaluate.js:643]:
```suggestion
  it('rejects tasks not allowed by the tasking algorithm', async () => {
```

It's really not allowed for the node _in that round_. Don't know if we need to be so specific. I think this is clearer and simpler

File: lib/tasker.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ */
+export async function getTaskKey (task, randomness) {
+  assert.strictEqual(typeof task, 'object', 'task must be an object')
+  assert.strictEqual(typeof task.cid, 'string', 'task.cid must be a string')
+  assert.strictEqual(typeof task.minerId, 'string', 'task.minerId must be a string')
+  assert.strictEqual(typeof randomness, 'string', 'randomness must be a string')
+
+  const data = [task.cid, task.minerId, randomness].join('\n')
+  const hash = await crypto.subtle.digest('sha-256', textEncoder.encode(data))
+  return BigInt('0x' + encodeHex(hash))
```
Comment (juliangruber) [inline lib/tasker.js:73]:
```suggestion
  return BigInt('0x' + Buffer.from(hash).toString('hex'))
```

This function is so simple, I think refactoring it adds to the cognitive load more than it reduces it

File: lib/tasker.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+/**
+ * @param {string} stationId
+ */
+export async function getStationKey (stationId) {
+  assert.strictEqual(typeof stationId, 'string', 'stationId must be a string')
+
+  const hash = await crypto.subtle.digest('sha-256', textEncoder.encode(stationId))
+  return BigInt('0x' + encodeHex(hash))
```
Comment (juliangruber) [inline lib/tasker.js:83]:
```suggestion
  return BigInt('0x' + Buffer.from(hash).toString('hex'))
```

File: lib/tasker.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+function getStationTaskDistance (stationKey, task) {
+  return stationKey ^ task.key
+}
+
+/**
+ * @param {ArrayBuffer} arrayBuffer
+ */
+function encodeHex (arrayBuffer) {
+  return Buffer.from(arrayBuffer).toString('hex')
+}
```
Comment (juliangruber) [inline lib/tasker.js:58]:
```suggestion
```

(see next comment)

File: lib/evaluate.js [hunk 5]
```diff
@@ -228,8 +230,27 @@ export const evaluate = async ({
  * @param {bigint} roundIndex
  * @param {import('./preprocess.js').Measurement[]} measurements
  * @param {import('./typings.js').RoundDetails} sparkRoundDetails
+ * @param {Pick<Console, 'log' | 'error'>} logger
  */
-export const runFraudDetection = async (roundIndex, measurements, sparkRoundDetails) => {
+export const runFraudDetection = async (roundIndex, measurements, sparkRoundDetails, logger) => {
+  const randomness = await getRandomnessForSparkRound(Number(sparkRoundDetails.startEpoch))
```
Comment (juliangruber) [inline lib/evaluate.js:236]: Let's watch this one in prod and add p-retry if it becomes a problem

**Review by juliangruber — CHANGES_REQUESTED** [review]
great work!

File: lib/tasker.js (hunk 2 above)
Comment (bajtos) [reply lib/tasker.js:73]: I see your point.

I am using `encodeHex` for consistency with the code in the spark checker repo.

How about introducing a function `hashToBigInt` that can use different implementations under the hood?

File: lib/evaluate.js (hunk 5 above)
Comment (bajtos) [reply lib/evaluate.js:236]: That's a great point! I did not consider that this external API service could fail 🙈 

Do you know if we have any alert that will post a Slack message when the round evaluation failed?

Comment (juliangruber) [reply lib/evaluate.js:236]: There is no such alert, it would be reported to Sentry :|

Comment (bajtos) [reply lib/evaluate.js:236]: Hmm, we are rather slow to act on Sentry errors.

Do you know if we have any alerts parsing the logs and posting a Slack message if a log message matches a given pattern? I think that would be the easiest solution.

File: lib/tasker.js (hunk 2 above)
Comment (bajtos) [reply lib/tasker.js:73]: SGTM!

File: lib/evaluate.js (hunk 5 above)
Comment (bajtos) [reply lib/evaluate.js:236]: https://www.papertrail.com/help/alerts/

Comment (bajtos) [reply lib/evaluate.js:236]: I configured Papertrail - Slack integration.

Comment (bajtos) [reply lib/evaluate.js:236]: See https://github.com/filecoin-station/spark-evaluate/pull/297

File: lib/tasker.js (hunk 2 above)
Comment (bajtos) [reply lib/tasker.js:75]: See cb7d18c

**Review by juliangruber — APPROVED** [review]
amazing!

fireworks

**Comment by juliangruber** [comment]
Looks like unit tests are failing

**Comment by bajtos** [comment]
Yeah, the integration test was failing because of a logical conflict with https://github.com/filecoin-station/spark-evaluate/pull/298.

**Comment by bajtos** [comment]
> ## Suspect Issues
> This pull request was deployed and Sentry observed the following issues:
> 
> * ‼️ **TypeError: fetch failed** `getRandomnessForSparkRound(drand-client)` [View Issue](https://space-meridian.sentry.io/issues/5689220767/?referrer=github-pr-bot)
> 
> Did you find this useful? React with a 👍 or 👎

See https://github.com/filecoin-station/spark-evaluate/issues/309

**Pull request merged**